src/fu-self-test.c

```diff
@@ -3176,6 +3178,162 @@ _plugin_device_register_cb(FuPlugin *plugin, FuDevice *device, gpointer user_dat
 	fu_plugin_runner_device_register(plugin, device);
 }
 
+static void
+fu_backend_usb_func(gconstpointer user_data)
+{
+	FuTest *self = (FuTest *)user_data;
+	gboolean ret;
+	FuDevice *device_tmp;
+	g_autofree gchar *gusb_emulate_fn = NULL;
+	g_autofree gchar *devicestr = NULL;
+	g_autoptr(FuBackend) backend = fu_usb_backend_new(self->ctx);
+	g_autoptr(FuDeviceLocker) locker = NULL;
+	g_autoptr(FuProgress) progress = fu_progress_new(G_STRLOC);
+	g_autoptr(GError) error = NULL;
+	g_autoptr(GPtrArray) devices = NULL;
+	g_autoptr(GPtrArray) possible_plugins = NULL;
+	g_autoptr(JsonParser) parser = json_parser_new();
+
+#if !G_USB_CHECK_VERSION(0, 4, 0)
+	g_test_skip("GUsb version too old");
+	return;
+#endif
+
+	/* generate the fake base64 chunks in the usb-devices.json file */
+	if (g_getenv("FWUPD_GENERATE_DS20") != NULL) {
+		fwupd_guid_t guid = {0x0};
+		guint32 fwupd_version = (1u << 16) | (8u << 8) | 5u;
+		g_autofree gchar *str2 = NULL;
+		g_autofree gchar *str = NULL;
+		g_autoptr(GByteArray) buf2 = g_byte_array_new();
+		g_autoptr(GByteArray) buf = g_byte_array_new();
+
+		/* bReserved */
+		fu_byte_array_append_uint8(buf, 0x0);
+
+		/* PlatformCapabilityUUID */
+		ret = fwupd_guid_from_string("010aec63-f574-52cd-9dda-2852550d94f0",
```

Comment:
wait, isn't it supposed to be `fwupd_guid_from_string("fwupd"`?

Reply:
fwupd_guid_from_string() takes a GUID and parses it into a `fwupd_guid_t` blob. Be we certainly could do `fwupd_guid_hash_string()` to get the former.

Reply:
`fwupdtool firmware-build` to the rescue.